dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NormalizeLoops.java

```diff
@@ -70,4 +116,59 @@ public void run(Function function) {
             counter++;
         }
     }
+
+    private void guaranteeSingleEntry(Function function) {
+        int labelCounter = 0;
+        for (Label label : function.getEvents(Label.class)) {
+
+            final List<CondJump> backJumps = label.getJumpSet().stream()
+                    .filter(j -> j.getLocalId() > label.getLocalId())
+                    .sorted()
+                    .toList();
+            if (backJumps.isEmpty()) {
+                continue;
+            }
+
+            int counter = 0;
+
+            final Label loopBegin = label;
+            final CondJump uniqueBackJump = backJumps.get(0);
+
+            final Register sourceReg = function.newRegister(String.format("__jumpedTo_%s#%s_From", label, labelCounter), types.getArchType());
+            final Local initJumpedFromOutside = EventFactory.newLocal(sourceReg, expressions.makeZero(types.getArchType()));
+            function.getEntry().insertAfter(initJumpedFromOutside);
+
+            final List<Label> loopBodyLabels = loopBegin.getSuccessor().getSuccessors().stream()
+                    .takeWhile(ev -> ev != uniqueBackJump)
+                    .filter(Label.class::isInstance).map(Label.class::cast)
+                    .toList();
+
+            for (Label l : loopBodyLabels) {
+                final List<CondJump> externalEntries = l.getJumpSet().stream()
+                        .filter(j -> j.getLocalId() < loopBegin.getLocalId())
+                        .toList();
+
+                for (CondJump fromOutside : externalEntries) {
+                    counter++;
+
+                    final Expression source = expressions.makeValue(BigInteger.valueOf(counter), types.getArchType());
+                    final Local jumpingFrom = EventFactory.newLocal(sourceReg, source);
+                    final CondJump jumpToHeader = EventFactory.newGoto(loopBegin);
+
+                    fromOutside.replaceBy(jumpingFrom);
+                    jumpingFrom.insertAfter(jumpToHeader);
+                    jumpToHeader.updateReferences(Map.of(jumpToHeader.getLabel(), loopBegin));
```

Comment:
Why don't you just retarget the external jump and insert the reg assignment before. As I said before, this is to preserve the metadata of the original jump (such as source location).
In fact, the `updateReferences` doesn't do anything because you call it on the newly generated jump.
```Java
fromOutside.updateReferences(Map.of(fromOutside.getLabel(), loopBegin));
fromOutside.getPredecessor().insertAfter(jumpingFrom);
```